### ISSUE TITLE
convert torch.jit.script model to ONNX get wrong result

### ISSUE THREAD

**Issue by MrRace** [pytorch/pytorch#88072]
### 🐛 Describe the bug

I want to convert  `torch.jit.script` model  to ONNX, however the inference of ONNX and raw pytorch is different. My code snippet：
```python
    import torch.nn as nn
    import torch
    import torch.nn.functional as F
    import numpy as np
    import onnxruntime as ort
    import pdb


    class test_net(nn.Module):
        def __init__(self,):
            super(test_net, self).__init__()
            self.model = nn.Conv3d(3,64,kernel_size=(1,3,3), stride=(2,1,2))
            self.relu = nn.ReLU()
            self.relu6 = nn.ReLU6()
            self.relu66 = nn.ReLU6()

        def forward(self, x):
            out1 = self.model(x)
            f_mean = torch.mean(out1) # -> ReduceMean
            out2 = torch.div(out1, f_mean)
            o1 = self.relu(out2)
            o2 = self.relu6(out2)
            o3 = self.relu66(out2)   # script模式下被优化掉了
            out = torch.cat([o1,o2,o3])

            return out

    # 模型构建和运行
    imgh, imgw = 24, 94
    net = test_net().eval() # 若存在batchnorm、dropout层则一定要eval() 使得BN层参数不更新
    dummy_input = torch.randn(1, 3, 3, imgh, imgw)# n c d h w
    torch_out = net.forward(dummy_input)# net(dummy_input)


    # export onnx
    dynamic_axes = {'input': {3: 'height', 4: 'width'}, 'output': {3: 'height', 4: 'width'}} # 配置动态分辨率
    onnx_pth = "test-conv-relu.onnx"

    
    # 传入 scriptModule
    net_script= torch.jit.script(test_net().eval()) # 生成一个 ScriptModule
    # 需要外加配置 example_outputs，用来获取输出的shape和dtype，无需运行模型
    torch.onnx.export(net_script, dummy_input, onnx_pth, input_names=['input'], output_names=['output'], opset_version=11, 
        dynamic_axes=dynamic_axes)
    # example_outputs=[torch_out]

    onnx_session = ort.InferenceSession(onnx_pth, providers=['CUDAExecutionProvider'])
    onnx_blob = dummy_input.data.numpy()
    onnx_out = onnx_session.run(None, {'input':onnx_blob})[0]

    print('mean diff = ', np.mean(onnx_out - torch_out.data.numpy()))
```
There is a significant difference between `onnx_out ` and  `torch_out`

PS:  It seems Pytorch 1.13  `torch.onnx.export` does not have `example_outputs`  parameter

### Versions

Pytorch Version: 1.13.0a0+08820cb

**Comment by justinchuby** [comment]
Hi! Could you compare the results with `torch.testing.assert_allclose` and share your results? I wonder what the diff is and if the shapes and dtypes etc. match.

**Comment by thiagocrepaldi** [comment]
On your example above, the pytorch model is running on `cpu` while on `onnxruntime` it is running on `cuda`. That alone could cause significant result differences.

try changing your model/input to be on cuda and use onnnxruntime on cuda as well, to make sure the same hardware is used in both versions

also, if `x` is not a tensor for your `forward(self, x)`, type annotate it as torch JIT script will assume it is torch.Tensor when annotation is not present.

Another suggestion is to use the jighest opset_version possible, such as 16 or 17. Newer versions are more likely to receive bug fixes/improvements than older/less used ones

Finally, report the actual output you get in both for us to better understand what is "significant different". We usually consider result match for `rtol=1e-3, atol=1e-6`

**Comment by thiagocrepaldi** [comment]
@justinchuby a runnable repro was filed. Why do we need the `needs reproduction` label?

**Comment by AllenTiTaiWang** [comment]
`needs reproduction` label on the contrary, is suggesting this issue has a repro for instructors, and anyone should repro first before giving instructions, according to the label definition. I guess @justinchuby is scanning this issue, and marks this for his future TODO.

**Comment by MrRace** [comment]
> On your example above, the pytorch model is running on `cpu` while on `onnxruntime` it is running on `cuda`. That alone could cause result differences (due to hardware differences). Try changing your pytorch model/input to be on cuda and use keep using onnnxruntime on cuda, to make sure the same hardware is used in both versions
> 
> Also, if `x` is not a tensor for your `forward(self, x)`, type annotate it as torch JIT script will assume it is torch.Tensor when annotation is not present.
> 
> Another suggestion is to use the latest opset_version when possible, such as 16 or 17. Newer versions are more likely to receive bug fixes/improvements than older/less used ones.
> 
> Finally, we usually consider result match for `rtol=1e-3, atol=1e-6`, check if that covers your difference (it doesnt seem to be the case)

@thiagocrepaldi  I check again, and find that if use `net_script= torch.jit.script(test_net().eval())` to creat ScriptModule the `mean diff =  15.77094`. When use `net_script= torch.jit.script(net)` which reuse the previous Model instance, the `mean diff =  -1.3363443e-05`. Why ?  it seems unreasonable.

**Comment by justinchuby** [comment]
Hopefully it works now.
